gr-digital/lib/header_format_crc.cc

```diff
@@ -47,10 +49,11 @@ bool header_format_crc::format(int nbytes_in,
     // for 12-bit representation?
     nbytes_in &= 0x0FFF;
 
-    d_crc_impl.reset();
-    d_crc_impl.process_bytes((void const*)&nbytes_in, 2);
-    d_crc_impl.process_bytes((void const*)&d_header_number, 2);
-    uint8_t crc = d_crc_impl();
+    unsigned char buffer[] = { (unsigned char)(nbytes_in & 0xFF),
```

Comment:
I think this is only doing the same thing as the previous code if the platform is little-endian. But I like this version better, as it does the same thing regardless of the endianness.

But then I wonder if the order in which the bytes of the `nbytes_in` and `d_header_number` fields end up in the output depends on the machine endianness. Looking at the code I guess it does depend. So we might end-up having something weird where the CRC isn't calculated over the bytes in the order they end up in the header but in some other order.

Perhaps it's good to take a closer look to this code and try to get a behaviour that is consistent regardless of the endianness and the CRC calculated using the same order as the data in the header?

Reply:
> I think this is only doing the same thing as the previous code if the platform is little-endian. But I like this version better, as it does the same thing regardless of the endianness.

I suspect the old code didn't work at all for big-endian platforms, because `nbytes_in` is an `int`, which is usually 4 bytes and therefore would have zeroes as its first two bytes.

Reply:
>But then I wonder if the order in which the bytes of the nbytes_in and d_header_number fields end up in the output depends on the machine endianness. Looking at the code I guess it does depend.

Does it? It looks to me like `add_field32` includes endianness conversion (via `volk_32u_byteswap`):

https://github.com/gnuradio/gnuradio/blob/6767475c9312520395cb888997d3348de32b1054/gr-digital/lib/header_buffer.cc#L72-L84

Reply:
I hadn't look at `add_field32` in detail, but now that I do, I note two things:
* `header_format_crc::format` always calls `add_field32` with `bs = true`, so the `byteswap` is never done.
* I don't really understand the intention of `x = x >> (32 - len);` after the byteswap. I think that if you feed the same number in `data` and use opposite values for `bs` in a little-endian and big-endian system, if you don't do this shift `x = x >> (32 - len);`, then you get the same data appended to `d_buffer`. But if you do the shift, you get different data appended to the buffer (unless `len == 32`, of course).